### PR TITLE
Fixed #28126 -- Added GistIndex to contrib.postgres.

### PR DESCRIPTION
**Pull request by atombrella**
Thanks to Marc Tamlyn for the initial patch.

Follow up on #8303 I added the `BtreeGistExtension` and support for `fillfactor` and `buffering` plus schema tests.

### CODE REVIEW

File: django/contrib/postgres/indexes.py [hunk 1]
```diff
@@ -57,3 +57,37 @@ def get_sql_create_template_values(self, model, schema_editor, using):
...
+        self.buffering = buffering
+        super().__init__(**kwargs)
+
+    def deconstruct(self):
+        path, args, kwargs = super().deconstruct()
+        kwargs['fillfactor'] = self.fillfactor
+        kwargs['buffering'] = self.buffering
+        return path, args, kwargs
+
+    def get_sql_create_template_values(self, model, schema_editor, using):
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: This needs an update following 831358f23d545b8dba017c6b26bd295ba9f6c17d.

File: django/contrib/postgres/indexes.py [hunk 2]
```diff
@@ -57,3 +57,37 @@ def get_sql_create_template_values(self, model, schema_editor, using):
 
     def create_sql(self, model, schema_editor):
         return super().create_sql(model, schema_editor, using=' USING gin')
+
+
+class GistIndex(Index):
+    suffix = 'gist'
+    # Allow an index name longer than 30 characters since the suffix is 4
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: I'd move this to a mixin to share with BrinIndex or drop the repetitive comment.

File: django/contrib/postgres/indexes.py [hunk 3]
```diff
@@ -57,3 +57,37 @@ def get_sql_create_template_values(self, model, schema_editor, using):
...
+    max_name_length = 31
+
+    def __init__(self, *, fillfactor=None, buffering=None, **kwargs):
+        self.fillfactor = fillfactor
+        self.buffering = buffering
+        super().__init__(**kwargs)
+
+    def deconstruct(self):
+        path, args, kwargs = super().deconstruct()
+        kwargs['fillfactor'] = self.fillfactor
```
Comment (timgraham) [inline django/contrib/postgres/indexes.py]: I think there's a mistake in `GinIndex.deconstruct()` also... these should be guarded with, e.g. `if self.fillfactor is not None` similar to logic in model fields' `deconstruct()`.

File: django/contrib/postgres/indexes.py (hunk 1 above)
Comment (atombrella) [inline django/contrib/postgres/indexes.py]: I think that'd be done by a rebase.

File: django/contrib/postgres/indexes.py (hunk 3 above)
Comment (atombrella) [inline django/contrib/postgres/indexes.py]: Same commit to fix all cases, including `BrinIndex`, or separate commits?

Comment (atombrella) [reply django/contrib/postgres/indexes.py]: I forgot the test when updating, but I just pushed it.

**Pull request merged**